### DISCUSSION TITLE
Move rounding of sensor states from backend to frontend

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#636] [General] [answered]
### Background
Rounding in core of sensor states is happening in many places in core. This is mostly done for the sole purpose of not having unsightly numbers with many decimals displayed in the frontend, Python's `round` function is called some 700 times in core, not including tests.

The rounding is hit-and-miss, some integrations do rounding and some don't. The state will sometimes be rounded by a base component or the state machine itself in certain conditions but mostly not.

### Problem
- We should not round the state in core to make it look nice when presented, that should be done by the frontend when needed
- Accumulating numbers which have been aggressively rounded for display purposes introduces large growing errors which causes the accumulated sum to be significantly wrong.

### Suggestion
- If the state is a valid number, and the entity's attributes are such that we would display a line chart (either unit of measurement or state class is set):
  - User can customize the number of decimals for a certain entity
  - If user has not customized, fall back to a new entity attribute `precision` set by integrations on the entity
  - If attribute `precision` is not defined, fall back to special cases for known device classes as needed, for example 1 decimal for temperatures in °C and no decimals for temperatures in °F
  - If the device class is not a special case, fall back to a user configurable default rounding which applies to all entities

#### Problems with the suggestion
* How to allow user to customize the rounding, should it be done by overriding the `precision` attribute as an entity customization, as a configuration of the sensor Lovelace card or in some other way?
* Do we want to introduce an entity attribute for this purpose, or should integrations not have a say at all?
  As an example, consider an integration providing sensors with energy price. The energy price is likely to be exact down to 1/100th of a cent, maybe it doesn't matter if the price is rounded to a lower precision when displayed.

**Comment by Jc2k** [comment]
Very much a fan of centralising rounding like this.

Do we want the integration to get a say at all? Personally i'd like all my sensors to be consistent, and one integration overriding the default would be annoying.

**Comment by pvizeli** [comment]
I think that is a fronted-driven issue that can not be solved with a simple attribute `precision`. And still, if we add a static attribute that going into the state machine and only can edit using YAML and the old overwrite functions. If that is really a big issue, we need full UI support to solve that and not only our old mechanics from the old state UI times.

I think we can assume a pretty close default rounding on UI based on device_class and otherwise with a default of 2. If that will not work for a user, they can use the Lovelace advance attribute to overwrite that precision. So we solve the issue for 98% - 99% of the user out of the box and allow the other 1% to have what they want.

It will be also not simple for the integration to assume the correct precision and I see mostly we have only maybe 5-8 integration which people complain about today and maybe can be solved better as just a dump attribute.

**Reply by emontnemery** [reply]
Right, so what you're saying is essentially this:
- User can customize the number of decimals for a certain entity
- ~If user has not customized, fall back to a new entity attribute precision set by integrations on the entity~
- If user has not customized, fall back to special cases for known device classes as needed, for example 1 decimal for temperatures in °C and no decimals for temperatures in °F
- If the device class is not a special case, fall back to a user configurable default rounding which applies to all entities

I think it's more or less the same as what @Jc2k is saying (but maybe for a different reason)

**Comment by thecode** [comment]
@emontnemery there is one problem with not rounding in core:
Currently in Shelly integration we round some of the values to reduce state changes, I agree with you users might prefer different precisions,  for example we round the voltage to 1 decimal and energy to 2 decimals.

The problem is that if we don't round at all the energy sensors which derives from `Wmin` and reported as `kWh` would have many state changes.
I agree that rounding may introduce errors, and again the energy is a good example where you don't want to round, but we should think of a solution that will fit both.

Ideally integration can report the raw unrounded value and the requested precision so that internal calculations can run on the raw value and state changes will only happen after rounding. I understand this might have more complicated changes in core, but on the other side does not require any frontend changes (except maybe if we want to allow user to customize the precision) since state would already be rounded.

**Reply by farmio** [reply]
Isn't that what significant change is for? https://developers.home-assistant.io/docs/core/platform/significant_change

**Comment by BradleyFord** [comment]
Now that we are getting more advanced in terms of standardizing units of
measure, should this be expanded to set the number of decimal places as
part of that configuration?
So we limit it in the backend to limit state changes.
Then the front end could still have a precision setting

**Comment by Kane610** [comment]
In regards to this topic; I right now limit temperature values from deconz to one decimal. Would there be any effect on users automations and such if I remove the decimal limit?

**Comment by allenporter** [comment]
Anecdotally, *nest* users are interested in seeing decimals rounded for display https://github.com/home-assistant/core/issues/55496 but i've been resisting adding rounding to core, hoping for a general solution for display.  I received a PR this morning to add rounding to the integration.

@emontnemery sounds like nest users would appreciate anything you push forward here.

**Reply by michaeldavie** [reply]
It's also strange that the `climate` device appears to round the temperature value, which leads to different values being stored and displayed by two entities (`climate` and `sensor`) for the same measurement.

**Reply by Jc2k** [reply]
> It's also strange that the `climate` device appears to round the temperature value, which leads to different values being stored and displayed by two entities (`climate` and `sensor`) for the same measurement.

I think the current rounding in climate is masking some weird behaviours too.

E.g. Climate rounding and conversion between F and C has been the source of a bunch of bugs for me in homekit_controller. The TL;DR is we have some devices that work natively in F, but their protocol is in C. If the HA instance is set to F that means you get a double conversion and precision loss. If we don't round those values then climate reports that temperature was reached and the thermostat is idle but might show .73246 (.3ish below temperature) and never actually "reach" the desired temperature.

This lead to cases where people reported the integration as broken, so we reverted the change to restore the original rounding and added a sensor entity for people wanting the raw data. Even though the actual code was working with and without the rounding it created a perception of brokenness.

I believe there was a similar issue in Ecobee too, though I'm not familiar with the specifics.

I don't think this would be needed if the rounding was a display time thing.

**Comment by balloob** [comment]
The current solution is not ideal, however anything that we come up with will also not be ideal and have other edge cases or issues. 

I don't want users to go around and being able to turn 100 knobs on how things should be rounded. It will make the frontend overly complicated and <1% will even ever use it (unless we pick horrible defaults).

**Comment by Mariusthvdb** [comment]
Unaware of this architecture discussion I posted this in #beta today: https://discord.com/channels/330944238910963714/427516175237382144/894611783594172467

suggesting a Lovelace card option on an entity(-row)

don’t want to say 15 decimals is horrible, but for the Frontend we really need to be able to easily adjust that.

**Comment by KNXBroker** [comment]
I have a lot of template sensors only for rounding. I never liked this, as the information is duplicated everywhere. Furthermore attributes have to be assigned manually and integration changes are not overtaken automatically. Having this option in the frontend would be highly appreciated!  I appreciate that you want to keep the UI clean, but I think this should be a basic feature in the frontend cards' submenus.

**Comment by thecode** [comment]
I still think that my main concern here was not answered: https://github.com/home-assistant/architecture/discussions/636#discussioncomment-1332635

If we remove rounding from the integration, an integration that has a voltage with millivolt resolution would cause high number of state updates, e.g a sensor that reports 230.053 will be constantly changing.

**Reply by elupus** [reply]
> I still think that my main concern here was not answered: [#636 (comment)](https://github.com/home-assistant/architecture/discussions/636#discussioncomment-1332635)
> 
> If we remove rounding from the integration, an integration that has a voltage with millivolt resolution would cause high number of state updates, e.g a sensor that reports 230.053 will be constantly changing.

Rounding doesn't always help there. The device could be jittering between 0.49 and 0.5. if you where to round to 0 decimal places, it would jitter between 1 and 0 as state change. You need to keep track of significant changes. Ie the difference in the change.

**Reply by michal-bartak** [reply]
what about to provide defaults per unit.
Such roundings should be picked up and applied by UI cards only.

Why per unit? because in general users wants it's UI to be consistent. However cards could have feature of overriding the defaults

Commenting on fraction of voltage changes frequently... it's not the core business. It's an integration which is responsible for making decision about what is feasible or not potentially giving option for throttling or any other method dropping unipmortant changes

**Reply by mvdwetering** [reply]
There is a feature called Significant Change Support (https://developers.home-assistant.io/docs/significant_change_index/) can that help for this case?

**Reply by jbouwh** [reply]
There is paradox in wanting fast UI updates and state changes that occur. Somehow all those states changes stored but mostly not always needed when showing a graph, that is only interesting for sensors with limited states. For number values like power we have statistics, so that lowers the need to keep a lot of regular logging. For who wants to look back for the details over a longer time using influxdb is a much better option, with better performance imo.
Perhaps we could consider if we could use that as a backend for recorder as an alternative to SQL.

**Comment by parautenbach** [comment]
Does it need to be backend or frontend? Can't both options be allowed? Generally speaking, making this a presentation issue is usually the best way to go, but I completely get the MQTT comments. Perhaps there should be some level of rounding for integrations allowed (measurements could be noisy and not significant).

**Reply by mmiller7** [reply]
> Does it need to be backend or frontend? Can't both options be allowed? Generally speaking, making this a presentation issue is usually the best way to go, but I completely get the MQTT comments. Perhaps there should be some level of rounding for integrations allowed (measurements could be noisy and not significant).

I think this is important.

There are some things that you do want to round on the "back-end" for example if I want to manually account for known significant figures in a computation or known accuracy I want to round it down to only the useful amount of digits.

At the same time, I may have a sensor that is accurate to say 2 decimals and want to preserve that, but for display purposes I only care about the whole number.

Suppose I have a battery sensor which reports its capacity such that I can compute the % charge to 2-3 decimals...do I really care if it's 75.05% or 75% on display?  Probably not.  But if I'm computing a small draw over a long time that 0.05% difference may be significant to projecting rate of change over a long interval.

Having a back-end "precision" and a separate front-end "display_precision" would allow greater flexibility.

**Comment by smizles** [comment]
> * How to allow user to customize the rounding, should it be done by overriding the `precision` attribute as an entity customization, as a configuration of the sensor Lovelace card or in some other way?

Could the user-defined default precision be specified in the UI > Configuration > General?  
![20211118-UI_General_Configuration-mu](https://user-images.githubusercontent.com/25838538/142380064-6003a1a1-1637-4d85-a7b7-36fdd08fefea.PNG)

**Reply by michal-bartak** [reply]
But precision should be set per unit at least. You cannot assume the same precision for temperature and currency. Even different units for the same measured values might require different precision (Bar vs PSi). So.. ok for global configuration as long as it covers all units (the list could be user extendable)

**Comment by lukepighetti** [comment]
It is common practice in frontend development to allow the backend to present any precision it likes, allowing the frontend engineer to specify format at the template level.

Jinja has formatting primitives described here: https://support.sendwithus.com/jinja/formatting_numbers/

It seems like we should be able to add a field to all cards that is `format` which exposes a value `value` which is passed through Jinja

```yaml
type: gauge
entity: sensor.indoor_temperature_mean
format: value|round
```

I don't believe there is a need to totally refactor/rework how precision is handled on the backend. In the case that it is a string, Jinja has tools for coercing types. This could also stop any further bikeshedding as nothing else need change.

**Comment by timdonovanuk** [comment]
Ended up here because the sensor card doesn't support any form of rounding and I was surprised. I don't see why precision would need to be set at a global level, nor does it need an architectural shift. HA just needs to ensure there is the option to provide frontend formatting of values in every core card (such as sensor).

**Comment by abdavid** [comment]
Ended up here like @timdonovanuk while trying to find out how to round simple sensor values in an elegant way. Current templating solutions are overkill. 

I wholeheartedly second the suggestion from @lukepighetti The norm is that the backend can be as detailed as it wants, while the frontend formats and presents data that makes sense for that individual domain. No need for changing the backend as I see it, but adding formatting possibilities to the core lovelace cards is what the community is asking for. Current solutions for this has large thresholds in terms of knowledge, and also introduces overhead and more complexity. 

Just my 2 cents.

**Reply by Jc2k** [reply]
You might be interested in the (work in progress) releases notes for next weeks release.

https://rc.home-assistant.io/blog/2023/01/25/release-20232/#sensor-precision

**Reply by Tuxie** [reply]
> "This is an true entity setting, which means that changing the precision, will actually affect your entity state. So, it applies not just to your dashboard, but everything that uses the entity, including automations."

Why, oh why??? That it so very backwards and it's phrased like it's a feature. So if I want to display a rounded value on the dashboard, I will also get staircase graphs for that value because precision was discarded in the database? Rounding is a _presentation_ thing only, data should always be stored with as much precision as possible.

**Reply by mmiller7** [reply]
> > "This is an true entity setting, which means that changing the precision, will actually affect your entity state. So, it applies not just to your dashboard, but everything that uses the entity, including automations."
> 
> Why, oh why??? That it so very backwards and it's phrased like it's a feature. So if I want to display a rounded value on the dashboard, I will also get staircase graphs for that value because precision was discarded in the database? Rounding is a _presentation_ thing only, data should always be stored with as much precision as possible.

Is this seriously how it will work?!

This is the absolute stupidest way to implement it and will do a lot worse than "staircase graphs" - it will ruin things that actually NEED the decimals to do computation.

When I have something like a battery % report and I'm computing an estimated time to empty with a statistics sensor it may only change by 0.1% per 15 minutes (or whatever interval I compute the statistics sensor over) but even 0.05% is very critical if you are estimating a runtime of 1 day or 30 days remaining at the current draw.  But on the main UI there's no reason to show more than the whole % number.

If changing what it shows changes the back-end values then a slowly changing graph like a large solar-battery bank discharging will end up showing 0 discharge rate for the 15 minute sample interval because the whole number isn't changing but every few hours, then it will show a really inaccurate rate when the whole number steps.

**Reply by michal-bartak** [reply]
I'll tell you one more thing if you are not aware about.
Currently, HA sensors values might be cast to integers or floating point domains. There is no way to cast to decimal. So rounding the sensor value doesn't prevent getting a mess in GUI if further operations are involved.

**Reply by lukepighetti** [reply]
Did they add front-end rounding as well? https://github.com/home-assistant/frontend/pull/15176

**Reply by melyux** [reply]
> Did they add front-end rounding as well? https://github.com/home-assistant/frontend/pull/15176

Looks like they reverted that seemingly great change... https://github.com/home-assistant/frontend/pull/15305. Why on earth would they do that

**Reply by emontnemery** [reply]
Rounding is now implemented in frontend instead: https://github.com/home-assistant/architecture/discussions/636#discussioncomment-4890575

**Reply by lukepighetti** [reply]
any insights as to how this made it through to begin with?

**Reply by frenck** [reply]
Sure! In both cases, the goal is to provide controls to rounding/precision to the end user AND remove the logic from all integrations that now do so themselves.

The previous approach accepted the native value and rounded in the state engine. As we talk about display, we have more display than just the frontend. For example, template output, Alexa, Google Assistant, and more places. Rouding at the state engine, has an advantage in those cases. Not just that, if lesser precision is requested, it would also mean lesser state changes, and lesser writes to database, history, and lesser state updates towards external parties (like Google, Alexa, Node-RED and more).

That worked fine in general; nevertheless, will give issues with statistics (like the energy dashboard), especially when switching between precisions. Discussing that, we looked at the possibility of adding native values that may be the recorder should use, or, as what is implemented now. Provide the ability for each output to figure out by themselves.

So, now the frontend handles it by itself, there is templating adjustment in place that allows users building template to leverage the rounding/display values configured. And for things like Alexa, it turned out they already round themselves.

../Frenck

**Comment by BradleyFord** [comment]
agree, it is quite odd to force the change to the raw data as well
(assuming that is the way it is actually done).

For the industrial data historian industry; precision can be configured for
the stored data, and separately where you use it in the UI you can set the
precision that you want to display.

**Comment by emontnemery** [comment, marked as answer]
A user configurable display precision has been implemented, and will be included in 2023.3. The solution allows the user to configure the number of decimal digits used for presentation of a sensor's value but leaves the actual sensor state unmodified:
- backend: https://github.com/home-assistant/core/pull/87330
- frontend: https://github.com/home-assistant/frontend/pull/15363

Note:
An earlier version of customized precision, which was included in the 2023.2 beta, rounded in the backend (instead of in frontend as is suggested in this architecture discussion):
- backend: https://github.com/home-assistant/core/pull/86074
- frontend: https://github.com/home-assistant/frontend/pull/15176
 
That version was removed the very last minute before the release of 2023.2 because rounding in the backend causes a lot of problems if the user configures the rounding because of presentation, as pointed out by @Tuxie and @mmiller7 in this thread, as well as elsewhere.

**Comment by michal-bartak** [comment]
BTW the rounding should be available per card (including specialized pages not built with cards, like energy, history etc).
The current solution allows setting rounding for a sensor, assuming that the user wants to show it this way regardless chosen UI item. Which obviously doesn't have to be true.

**Reply by frenck** [reply]
If a custom card overrides or ignores the property, that is up to the card itself.

If you'd like such functionality in the card provided by Core, I suggest raising a feature request instead, as that is not really architectural-related.

**Reply by michal-bartak** [reply]
I don't this approach is a good idea. There are reasons like consistency in user experience when using different cards.
Also, such an option has to override the entity rounding setting. Thus, it's not an FE case only (unless ignoring such rounding is implemented yet).
IMO it should be a platform standard, independent of FRs multiplicity.

**Reply by frenck** [reply]
> I don't this approach is a good idea. 

There is no other approach. This architectural discussion is closed and shipped.

> IMO it should be a platform standard, independent of FRs multiplicity.

It is. The user can provide its global rounding preference on an entity. Override on any point can happen as the original value remains available.
